Upstream patch for: -O2 miscompiles with -mtune=nehalem or corei7

This caused the GSL testsuite failure at http://trac.sagemath.org/ticket/17208
Upstream bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=63661
Patch is taken from PR: https://gcc.gnu.org/viewcvs/gcc?view=revision&revision=217783
Not yet in gcc 4.9.2, already in gcc trunk.

--- trunk/gcc/ira.c	2014/11/19 15:09:29	217782
+++ trunk/gcc/ira.c	2014/11/19 15:15:51	217783
@@ -5263,7 +5263,18 @@
 	      ira_allocno_iterator ai;
 
 	      FOR_EACH_ALLOCNO (a, ai)
-		ALLOCNO_REGNO (a) = REGNO (ALLOCNO_EMIT_DATA (a)->reg);
+                {
+                  int old_regno = ALLOCNO_REGNO (a);
+                  int new_regno = REGNO (ALLOCNO_EMIT_DATA (a)->reg);
+
+                  ALLOCNO_REGNO (a) = new_regno;
+
+                  if (old_regno != new_regno)
+                    setup_reg_classes (new_regno, reg_preferred_class (old_regno),
+                                       reg_alternate_class (old_regno),
+                                       reg_allocno_class (old_regno));
+                }
+
 	    }
 	  else
 	    {
